File: packages/edge/src/services/environment-registry.ts
/**
 * EnvironmentRegistry - Manages organization/project/environment tree for Edge
 *
 * Responsibilities:
 * - Fetch complete org/project/env tree on startup via internal API
 * - Maintain in-memory tree for fast lookups
 * - Listen to Redis PubSub events for real-time tree synchronization
 *   (environment.created, environment.deleted, project.created, project.deleted, etc.)
 */

import axios from 'axios';
import Redis from 'ioredis';
import { config } from '../config/env';
import { createLogger } from '../config/logger';

const logger = createLogger('EnvironmentRegistry');

export interface EnvironmentNode {
  id: string;
  name: string;
  displayName?: string;
  environmentType?: string;
}

export interface ProjectNode {
  id: string;
  projectName: string;
  displayName?: string;
  environments: EnvironmentNode[];
}

export interface OrgNode {
  id: string;
  orgName: string;
  displayName?: string;
  projects: ProjectNode[];
}

/**
 * EnvironmentRegistry - Singleton service
 */
class EnvironmentRegistry {
  private tree: OrgNode[] = [];
  private subscriber: Redis | null = null;
  private initialized = false;
  private readonly CHANNEL_PREFIX = 'gatrix-sdk-events';
  private treeChangeCallbacks: Array<() => void> = [];

  // Lookup maps for fast access
  private envMap: Map<string, { orgId: string; projectId: string }> = new Map();
  private projectMap: Map<string, { orgId: string }> = new Map();

  /**
   * Initialize the registry
   */
  async initialize(): Promise<void> {
    if (this.initialized) {
      logger.warn('Already initialized');
      return;
    }

    logger.info('Initializing environment registry...');

    // Fetch initial tree
    await this.fetchTree();

    // Subscribe to environment lifecycle events
    await this.subscribeToEvents();

    this.initialized = true;
    logger.info('Initialized', {
      orgs: this.tree.length,
      projects: this.getAllProjects().length,
      environments: this.getAllEnvironmentIds().length,
    });
  }

  /**
   * Shutdown
   */
  async shutdown(): Promise<void> {
    if (this.subscriber) {
      try {
        await this.subscriber.punsubscribe(`${this.CHANNEL_PREFIX}:*`);
        await this.subscriber.quit();
        this.subscriber = null;
      } catch (error) {
        logger.error('Error during shutdown:', error);
      }
    }
    this.tree = [];
    this.envMap.clear();
    this.projectMap.clear();
    this.initialized = false;
    logger.info('Shutdown complete');
  }

  /**
   * Fetch the complete org/project/env tree from backend
   */
  async fetchTree(): Promise<void> {
    try {
      const response = await axios.get(
        `${config.gatrixUrl}/api/v1/server/internal/environment-tree`,
        {
          headers: {
            'x-api-token': config.apiToken,
            'x-application-name': config.applicationName,
          },
          timeout: 10000,
        }
      );

      if (response.data?.success && response.data?.data?.organisations) {
        this.tree = response.data.data.organisations;
        this.rebuildMaps();

        logger.info('Environment tree fetched', {
          orgs: this.tree.length,
          projects: this.getAllProjects().length,
          environments: this.getAllEnvironmentIds().length,
        });

        // Notify tree change listeners
        for (const cb of this.treeChangeCallbacks) {
          try {
            cb();
          } catch (e) {
            logger.error('Tree change callback error:', e);
          }
        }
      } else {
        logger.error('Invalid response from environment-tree API:', response.data);
      }
    } catch (error: any) {
      logger.error('Failed to fetch environment tree:', error.message);
      throw error;
    }
  }

  /**
   * Rebuild lookup maps from tree
   */
  private rebuildMaps(): void {
    this.envMap.clear();
    this.projectMap.clear();

    for (const org of this.tree) {
      for (const project of org.projects) {
        this.projectMap.set(project.id, { orgId: org.id });
        for (const env of project.environments) {
          this.envMap.set(env.id, { orgId: org.id, projectId: project.id });
        }
      }
    }
  }

  /**
   * Subscribe to lifecycle events
   */
  private async subscribeToEvents(): Promise<void> {
    try {
      this.subscriber = new Redis({
        host: config.redis.host,
        port: config.redis.port,
        password: config.redis.password || undefined,
        db: config.redis.db,
        lazyConnect: true,
        maxRetriesPerRequest: null,
        autoResubscribe: true,
      });

      await this.subscriber.connect();

      this.subscriber.on('pmessage', (_pattern: string, _channel: string, message: string) => {
        this.handleEvent(message);
      });

      await this.subscriber.psubscribe(`${this.CHANNEL_PREFIX}:*`);

      logger.info(`Subscribed to Redis pattern: ${this.CHANNEL_PREFIX}:*`);
    } catch (error: any) {
      logger.error('Failed to subscribe to events:', error.message);
      // Continue without real-time updates
    }
  }

  /**
   * Handle incoming lifecycle events
   */
  private handleEvent(message: string): void {
    try {
      const event = JSON.parse(message);
      const type = event.type;

      // Only handle environment/project/org lifecycle events
      if (
        type === 'environment.created' ||
        type === 'environment.deleted' ||
        type === 'project.created' ||
        type === 'project.deleted' ||
        type === 'org.created' ||
        type === 'org.deleted'
      ) {
        logger.info(`Received lifecycle event: ${type}`, { data: event.data });

        // Full tree refetch for simplicity and reliability
        this.fetchTree().catch((err) => {
          logger.error('Failed to refetch tree after event:', err.message);
        });
      }
    } catch (error: any) {
      logger.error('Failed to parse event:', error.message);
    }
  }

  // ==================== Query Methods ====================

  /**
   * Get the full tree
   */
  getTree(): OrgNode[] {
    return this.tree;
  }

  /**
   * Get all environment IDs
   */
  getAllEnvironmentIds(): string[] {
    return Array.from(this.envMap.keys());
  }

  /**
   * Get all project IDs
   */
  getAllProjects(): ProjectNode[] {
    return this.tree.flatMap((org) => org.projects);
  }

  /**
   * Get org/project for a given environment ID
   */
  getEnvironmentContext(environmentId: string): { orgId: string; projectId: string } | undefined {
    return this.envMap.get(environmentId);
  }

  /**
   * Get org for a given project ID
   */
  getProjectContext(projectId: string): { orgId: string } | undefined {
    return this.projectMap.get(projectId);
  }

  /**
   * Get environments for a specific project
   */
  getEnvironmentsByProject(projectId: string): EnvironmentNode[] {
    for (const org of this.tree) {
      for (const project of org.projects) {
        if (project.id === projectId) {
          return project.environments;
        }
      }
    }
    return [];
  }

  /**
   * Check if an environment ID exists
   */
  hasEnvironment(environmentId: string): boolean {
    return this.envMap.has(environmentId);
  }

  /**
   * Check if initialized
   */
  isInitialized(): boolean {
    return this.initialized;
  }

  /**
   * Force refresh
   */
  async refresh(): Promise<void> {
    await this.fetchTree();
  }

  /**
   * Register callback for tree changes.
   * Returns an unsubscribe function.
   */
  onTreeChanged(callback: () => void): () => void {
    this.treeChangeCallbacks.push(callback);
    return () => {
      this.treeChangeCallbacks = this.treeChangeCallbacks.filter((cb) => cb !== callback);
    };
  }
}

export const environmentRegistry = new EnvironmentRegistry();

File: packages/edge-rust/src/services/environment_registry.rs
// Gatrix Edge Rust - Environment Registry
// Manages organization/project/environment tree for Edge

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use log::{info, warn, error};
use serde::Deserialize;

use crate::config::EdgeConfig;

/// Environment node in the tree
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EnvironmentNode {
  pub id: String,
  pub name: String,
  pub display_name: Option<String>,
  pub environment_type: Option<String>,
}

/// Project node in the tree
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectNode {
  pub id: String,
  pub project_name: String,
  pub display_name: Option<String>,
  pub environments: Vec<EnvironmentNode>,
}

/// Organization node in the tree
#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OrgNode {
  pub id: String,
  pub org_name: String,
  pub display_name: Option<String>,
  pub projects: Vec<ProjectNode>,
}

/// Environment context (org + project for a given environment)
#[derive(Debug, Clone)]
pub struct EnvironmentContext {
  pub org_id: String,
  pub project_id: String,
}

/// Environment name resolution result
#[derive(Debug, Clone)]
pub struct EnvNameEntry {
  pub org_id: String,
  pub project_id: String,
  pub env_id: String,
}

/// Environment Registry service
pub struct EnvironmentRegistry {
  tree: Arc<RwLock<Vec<OrgNode>>>,
  env_map: Arc<RwLock<HashMap<String, EnvironmentContext>>>,
  env_name_map: Arc<RwLock<HashMap<String, EnvNameEntry>>>,
  initialized: Arc<RwLock<bool>>,
  config: EdgeConfig,
}

impl EnvironmentRegistry {
  pub fn new(config: EdgeConfig) -> Self {
    Self {
      tree: Arc::new(RwLock::new(Vec::new())),
      env_map: Arc::new(RwLock::new(HashMap::new())),
      env_name_map: Arc::new(RwLock::new(HashMap::new())),
      initialized: Arc::new(RwLock::new(false)),
      config,
    }
  }

  /// Initialize the registry
  pub async fn initialize(&self) -> Result<(), String> {
    let initialized = *self.initialized.read().await;
    if initialized {
      warn!("EnvironmentRegistry: Already initialized");
      return Ok(());
    }

    info!("EnvironmentRegistry: Initializing...");
    self.fetch_tree().await?;
    self.subscribe_to_events().await;

    let mut init = self.initialized.write().await;
    *init = true;

    let env_count = self.env_map.read().await.len();
    info!("EnvironmentRegistry: Initialized with {} environments", env_count);
    Ok(())
  }

  /// Fetch the complete org/project/env tree from backend
  pub async fn fetch_tree(&self) -> Result<(), String> {
    let url = format!(
      "{}/api/v1/server/internal/environment-tree",
      self.config.gatrix_url
    );

    let client = reqwest::Client::new();
    let response = client
      .get(&url)
      .header("x-api-token", &self.config.api_token)
      .header("x-application-name", &self.config.app_name)
      .timeout(std::time::Duration::from_secs(10))
      .send()
      .await
      .map_err(|e| format!("Failed to fetch environment tree: {}", e))?;

    let body: serde_json::Value = response
      .json()
      .await
      .map_err(|e| format!("Failed to parse tree response: {}", e))?;

    if body.get("success").and_then(|v| v.as_bool()) == Some(true) {
      if let Some(orgs_value) = body.get("data").and_then(|d| d.get("organisations")) {
        let orgs: Vec<OrgNode> = serde_json::from_value(orgs_value.clone())
          .map_err(|e| format!("Failed to deserialize tree: {}", e))?;

        self.rebuild_maps(&orgs).await;

        let mut tree = self.tree.write().await;
        *tree = orgs;

        let env_count = self.env_map.read().await.len();
        info!("EnvironmentRegistry: Tree fetched with {} environments", env_count);
        return Ok(());
      }
    }

    Err("Invalid response from environment-tree API".to_string())
  }

  /// Rebuild lookup maps from tree
  async fn rebuild_maps(&self, orgs: &[OrgNode]) {
    let mut env_map = self.env_map.write().await;
    let mut env_name_map = self.env_name_map.write().await;

    env_map.clear();
    env_name_map.clear();

    for org in orgs {
      for project in &org.projects {
        for env in &project.environments {
          env_map.insert(
            env.id.clone(),
            EnvironmentContext {
              org_id: org.id.clone(),
              project_id: project.id.clone(),
            },
          );
          env_name_map.insert(
            env.name.clone(),
            EnvNameEntry {
              org_id: org.id.clone(),
              project_id: project.id.clone(),
              env_id: env.id.clone(),
            },
          );
        }
      }
    }
  }

  /// Subscribe to lifecycle events via Redis PubSub
  async fn subscribe_to_events(&self) {
    let redis_host = self.config.redis.host.clone();
    let redis_port = self.config.redis.port;
    let redis_password = self.config.redis.password.clone();
    let redis_db = self.config.redis.db;
    let tree = Arc::clone(&self.tree);
    let env_map = Arc::clone(&self.env_map);
    let env_name_map = Arc::clone(&self.env_name_map);
    let config = self.config.clone();

    tokio::spawn(async move {
      let url = if let Some(password) = &redis_password {
        format!(
          "redis://:{}@{}:{}/{}",
          password, redis_host, redis_port, redis_db
        )
      } else {
        format!("redis://{}:{}/{}", redis_host, redis_port, redis_db)
      };

      let client = match redis::Client::open(url.as_str()) {
        Ok(c) => c,
        Err(e) => {
          warn!("EnvironmentRegistry: Failed to connect to Redis: {}", e);
          return;
        }
      };

      let mut pubsub = match client.get_async_pubsub().await {
        Ok(ps) => ps,
        Err(e) => {
          warn!("EnvironmentRegistry: Failed to get PubSub: {}", e);
          return;
        }
      };

      if let Err(e) = pubsub.psubscribe("gatrix-sdk-events:*").await {
        warn!("EnvironmentRegistry: Failed to subscribe: {}", e);
        return;
      }

      info!("EnvironmentRegistry: Subscribed to Redis pattern: gatrix-sdk-events:*");

      use futures_util::StreamExt;
      let mut stream = pubsub.on_message();

      while let Some(msg) = stream.next().await {
        let payload: String = match msg.get_payload() {
          Ok(p) => p,
          Err(_) => continue,
        };

        let event: serde_json::Value = match serde_json::from_str(&payload) {
          Ok(v) => v,
          Err(_) => continue,
        };

        let event_type = event
          .get("type")
          .and_then(|v| v.as_str())
          .unwrap_or("");

        // Only handle lifecycle events
        let lifecycle_events = [
          "environment.created",
          "environment.deleted",
          "project.created",
          "project.deleted",
          "org.created",
          "org.deleted",
        ];

        if !lifecycle_events.contains(&event_type) {
          continue;
        }

        info!("EnvironmentRegistry: Received lifecycle event: {}", event_type);

        // Refetch tree
        let fetch_url = format!(
          "{}/api/v1/server/internal/environment-tree",
          config.gatrix_url
        );

        let http_client = reqwest::Client::new();
        match http_client
          .get(&fetch_url)
          .header("x-api-token", &config.api_token)
          .header("x-application-name", &config.app_name)
          .timeout(std::time::Duration::from_secs(10))
          .send()
          .await
        {
          Ok(resp) => {
            if let Ok(body) = resp.json::<serde_json::Value>().await {
              if body.get("success").and_then(|v| v.as_bool()) == Some(true) {
                if let Some(orgs_value) =
                  body.get("data").and_then(|d| d.get("organisations"))
                {
                  if let Ok(orgs) =
                    serde_json::from_value::<Vec<OrgNode>>(orgs_value.clone())
                  {
                    // Rebuild maps
                    let mut em = env_map.write().await;
                    let mut enm = env_name_map.write().await;
                    em.clear();
                    enm.clear();

                    for org in &orgs {
                      for project in &org.projects {
                        for env in &project.environments {
                          em.insert(
                            env.id.clone(),
                            EnvironmentContext {
                              org_id: org.id.clone(),
                              project_id: project.id.clone(),
                            },
                          );
                          enm.insert(
                            env.name.clone(),
                            EnvNameEntry {
                              org_id: org.id.clone(),
                              project_id: project.id.clone(),
                              env_id: env.id.clone(),
                            },
                          );
                        }
                      }
                    }

                    let mut t = tree.write().await;
                    *t = orgs;

                    info!(
                      "EnvironmentRegistry: Tree refetched with {} environments",
                      em.len()
                    );
                  }
                }
              }
            }
          }
          Err(e) => {
            error!("EnvironmentRegistry: Failed to refetch tree: {}", e);
          }
        }
      }
    });
  }

  /// Resolve environment name or ID to actual environment ID
  pub async fn resolve_environment_id(&self, name_or_id: &str) -> Option<String> {
    // Try by name first
    let env_name_map = self.env_name_map.read().await;
    if let Some(entry) = env_name_map.get(name_or_id) {
      return Some(entry.env_id.clone());
    }
    drop(env_name_map);

    // Try by ID
    let env_map = self.env_map.read().await;
    if env_map.contains_key(name_or_id) {
      return Some(name_or_id.to_string());
    }

    None
  }

  /// Check if an environment ID exists
  pub async fn has_environment(&self, environment_id: &str) -> bool {
    self.env_map.read().await.contains_key(environment_id)
  }

  /// Get all environment IDs
  pub async fn get_all_environment_ids(&self) -> Vec<String> {
    self.env_map.read().await.keys().cloned().collect()
  }

  /// Get the tree
  pub async fn get_tree(&self) -> Vec<OrgNode> {
    self.tree.read().await.clone()
  }

  /// Check if initialized
  pub async fn is_initialized(&self) -> bool {
    *self.initialized.read().await
  }
}
